packages/phone-number-privacy/signer/src/key-management/key-provider-base.ts

```diff
@@ -1,28 +1,58 @@
+import { KeyName } from '@celo/phone-number-privacy-common'
+import config from '../config'
+export interface Key {
+  name: KeyName
+  version: number
+}
 export interface KeyProvider {
-  fetchPrivateKeyFromStore: () => Promise<void>
-  getPrivateKey: () => string
+  fetchPrivateKeyFromStore: (key: Key) => Promise<void>
+  getPrivateKey: (key: Key) => string
+  getPrivateKeyOrFetchFromStore: (key: Key) => Promise<string>
 }
 
 const PRIVATE_KEY_SIZE = 72
 
 export abstract class KeyProviderBase implements KeyProvider {
-  protected privateKey: string | null = null
+  protected privateKeys: Map<Key, string> = new Map()
 
-  public getPrivateKey() {
-    if (!this.privateKey) {
-      throw new Error('Private key is empty, provider not properly initialized')
+  public getPrivateKey(key: Key) {
+    const privateKey = this.privateKeys.get(key)
+    if (!privateKey) {
+      throw new Error(`Private key is unavailable: ${JSON.stringify(key)}`)
     }
+    return privateKey
+  }
 
-    return this.privateKey
+  public async getPrivateKeyOrFetchFromStore(key: Key): Promise<string> {
+    if (key.version < 1 || key.version > 10) {
+      throw new Error('Invalid private key version')
+    }
+    try {
+      return this.getPrivateKey(key)
+    } catch {
+      await this.fetchPrivateKeyFromStore(key)
+      return this.getPrivateKey(key)
+    }
   }
 
-  public abstract fetchPrivateKeyFromStore(): Promise<void>
+  public abstract fetchPrivateKeyFromStore(key: Key): Promise<void>
 
-  protected setPrivateKey(key: string) {
-    key = key ? key.trim() : ''
-    if (key.length !== PRIVATE_KEY_SIZE) {
+  protected setPrivateKey(key: Key, privateKey: string) {
+    privateKey = privateKey ? privateKey.trim() : ''
+    if (privateKey.length !== PRIVATE_KEY_SIZE) {
       throw new Error('Invalid private key')
     }
-    this.privateKey = key
+    this.privateKeys.set(key, privateKey)
+  }
+
+  protected getCustomKeyName(key: Key) {
+    switch (key.name) {
+      case KeyName.phoneNumberPrivacy:
+        return config.keystore.keys.phoneNumberPrivacy.name || key.name
+      case KeyName.domains:
+        return config.keystore.keys.domains.name || key.name
```

Comment:
Using the `??` gives 10 points to Gryffindor
```suggestion
        return config.keystore.keys.phoneNumberPrivacy.name ?? key.name
      case KeyName.domains:
        return config.keystore.keys.domains.name ?? key.name
```

Reply:
I think this might also be one of the few places where || is appropriate over ??. My reasoning here is that the env variables that we use to set the config values could be empty strings, in which case we would want to override that value with key.name. In general, I think we'd want to override any falsy value here but I may be missing something.